### PR TITLE
refactor(core): own identifiers and prototype composition

### PR DESCRIPTION
**Pull request by paulfalgout**
Related #241

## Linked issue

- Related #241

## What

- Add one owned identifier generator and route every Marionette constructor and event-listener ID through its shared counter.
- Replace Object and Application prototype composition with the existing safe own-assignment helper, removing both modules' complete Underscore imports.
- Document opaque Marionette-owned IDs, add focused ID/composition/lifecycle coverage, and regenerate affected distributions.

## Why

Migrating Object and Application alone would split the counter while other Marionette types still used Underscore, allowing equal custom prefixes to collide. This PR therefore moves every internal ID consumer atomically, preserving cross-type uniqueness while keeping the composition cleanup limited to Object and Application.

## Agent-development failure addressed

- Gives agents one local, tested source of truth for all framework-generated IDs instead of an implicit counter shared with an external utility library.
- Makes IDs explicitly opaque and framework-owned so agents do not infer numeric sequencing from unrelated Underscore or Backbone activity.
- Makes Object/Application mixin precedence, safe-key handling, descriptors, constructor order, and initialize arguments directly verifiable.

## Public behavior contract

- Canonical behavior after this PR: every Marionette-generated constructor and listener ID uses one monotonically increasing internal counter; equal custom prefixes remain unique across Marionette types; prefixes use JavaScript string coercion and failed coercion still consumes the attempted sequence number.
- Marionette IDs are opaque. Calls to external Underscore or Backbone ID generators no longer advance Marionette's internal sequence, and Marionette no longer advances theirs.
- Object and Application preserve constructor lifecycle order, option merging, radio/region initialization, static `extend`, mixin precedence, and initialize receiver/arguments.
- Superseded behavior removed: Object/Application direct Underscore `extend` and `uniqueId`, plus all other production Underscore `uniqueId` calls.
- Compatibility exception and removal condition: none; external counter coupling was an undocumented implementation detail and is intentionally removed for v5.

## Runtime cost

- [ ] Static or documentation only
- [ ] Development or test only; excluded from production entrypoints
- [x] Changes an existing production path
- [ ] Adds an explicitly opt-in runtime path

Measured impact or explanation:

- Bundle: the size/resource contract passes; direct production Underscore importers drop from 10 to 8. Brotli deltas versus exact base are ESM +33 B, CJS +39 B, UMD +26 B, and minified +12 B; cumulative measured artifacts grow 110 B.
- Startup/hot path: one module-scoped numeric counter replaces the Underscore counter path; each allocation performs one increment and one string concatenation.
- Allocations/retention: no retained per-instance state beyond the existing ID string.

## Scope

- Affects Marionette-generated IDs in Object, Application, Behavior, View, CollectionView, Region, and event-listener bookkeeping, plus Object/Application prototype composition.
- Does not remove other Underscore helpers from Behavior, View/Region, CollectionView, or Events; does not rename or remove MnObject/Application; and does not change regions, radio, start events, or inheritance APIs.

## Deployment Impact

- No website, package, tag, release, or deployment is performed by this PR.
- No consumer redeploy is required beyond adopting a future Marionette package containing this change.

## Validation

- [x] Tests cover the acceptance criteria and relevant edge cases.
- [x] Docs, examples, declarations, diagnostics, and rule metadata agree where relevant.
- [x] No private consumer, private fixture, or undocumented context is required.
- [x] I ran the commands listed below and am reporting their actual results.

Commands:

```sh
npm ci --no-audit --no-fund
npm exec vitest -- run test/unit/utils/unique-id.spec.js test/unit/object.spec.js test/unit/application.spec.js test/unit/object-application-composition.spec.js test/unit/behavior.spec.js test/unit/region.spec.js test/unit/view.spec.js test/unit/collection-view/collection-view.spec.js test/unit/mixins/events.spec.js --reporter=dot
npm exec vitest -- run test/unit/utils/unique-id.spec.js test/unit/object.spec.js test/unit/application.spec.js test/unit/object-application-composition.spec.js --reporter=dot
npm run lint:ci
git diff --check
npm test -- --reporter=dot
npm run build
npm run test:source
npm run test:dist
npm run test:fixtures
npm run docs:check
npm run check:browser-profile
npm run check:diagnostics
npm run test:performance
npm run size
npm run performance:timing
npm run check:dist
```

Results: broad focused tests passed 298/298; final exact coverage passed 32/32; full tests passed 1,508/1,508 across 81 files; source, distribution, packed CJS/ESM/types/Vite fixtures, documentation, browser profile, diagnostics, lint, build, size/resource, 92 performance-contract, hosted timing, post-commit distribution, and diff validation passed. Independent exact-head review passed 32/32 across four files with no findings.

## Package and browser impact

- Entry points or install fixtures affected: main ESM, CJS, UMD, and minified artifacts are regenerated; packed CJS/ESM/type/Vite fixtures pass.
- Browser contracts affected: none; the helper uses baseline numeric increment and string coercion only.
- Production/dev/test module-graph impact: Object and Application stop importing Underscore; every production ID call stops using it; remaining helper imports stay in their owning modules.

## Rollback or deprecation

- Revert this single commit and regenerate the main distribution artifacts; no data migration is required because generated IDs are runtime-local and documented as opaque.

## Review notes

- The atomic ID migration is intentional: a partial counter split could collide when migrated and unmigrated classes use the same custom prefix.
- Object/Application use `assignOwn`, matching Underscore `extend`'s own-enumerable string-key source contract while safely representing an own `__proto__` property and excluding inherited pollution.

---
## Summary by cubic
Owns all Marionette-generated IDs in one internal counter (`utils/unique-id.js`), so equal custom prefixes stay unique across types and external Underscore/Backbone calls no longer shift the sequence. Replaces Object and Application prototype composition with the existing `assignOwn` helper, preserving lifecycle and mixin behavior. Related to #241.

**Behavior changes**
- IDs are now framework-owned and documented as opaque; external `uniqueId` calls no longer affect Marionette's sequence or vice versa.
- Object and Application preserve constructor order, option merging, radio/region initialization, static `extend`, and mixin precedence.

**Migration**
- No consumer action is required beyond adopting the next Marionette package; rollback is reverting this change and rebuilding the dist artifacts.

<sup>Written for commit 1e5a81701129d190dbb3ff3db15482bd6a9b5bf7. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/marionettejs/marionette/pull/285?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by paulfalgout** [comment]
## Code-review approval — exact head `0111d100`

**Approved.** CI green, +110 B, 16,852 remaining. Production Underscore importers **10 → 8** — two cleared.

The generator is character-equivalent to `_.uniqueId`:

```js
let idCounter = 0;
const id = `${++idCounter}`;
return prefix ? prefix + id : id;
```

Same counter origin so the first ID is still `1`, same string coercion, same falsy-prefix branch — `uniqueId('')` and `uniqueId(0)` both return the bare number, as before.

**The atomicity argument in your description is the right call and worth affirming.** With a shared counter, two Marionette types using the same `cidPrefix` can never collide because every allocation advances one sequence. Migrating `object.js` and `application.js` alone would have left `view-region.js` and the event bookkeeping on Underscore's counter, and a custom prefix shared across that boundary would then produce duplicate `cid` values — which are used as registry keys. That's a subtle, data-dependent failure and doing all consumers in one PR is what avoids it.

**I was going to raise the flip side of that, and you've already documented it**, which I'd rather note than restate as a finding. Marionette IDs previously drew from the same Underscore counter as any application `_.uniqueId` call, so the two namespaces were incidentally disjoint; they now advance independently. The reference says the sequence *"is not coordinated with IDs generated by Underscore or Backbone"*, and the migration row states both directions plus the actionable guidance — don't parse suffixes, don't depend on allocation order, don't coordinate through `_.uniqueId`. That covers the practical case without inviting readers to reason about a collision scenario that only matters if those strings are used as shared keys.

The `_extend` → `assignOwn` swaps on the `Object` and `Application` prototypes are the same substitution reviewed on #280: identical reach over own enumerable keys, with inherited-enumerable copying (i.e. `Object.prototype` pollution) dropped.

Still source-only review; `node_modules` remains empty.

<sub>Reviewed with Claude Code.</sub>

File: test/unit/object-application-composition.spec.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    const protoValue = { safe: true };
+    mixins.forEach(mixin => {
+      const pollutedPrototype = {};
+      Object.defineProperty(pollutedPrototype, 'inheritedPollution', {
+        enumerable: true,
+        get() {
+          throw new Error('inherited pollution was read');
+        }
+      });
+      Object.setPrototypeOf(mixin, pollutedPrototype);
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply test/unit/object-application-composition.spec.js]: Handled in 1e5a817 by registering cleanup after each successful mutation and restoring all registered state in reverse order across setup and import failures.

**Comment by paulfalgout** [comment]
**Approval carries** to `1e5a8170` (the SHA I fetched). Tests-only — `dist/`, `mixins/`, `modules/`, `utils/`, and `docs/` are byte-identical to `0111d100`. CI green.

The restructuring is a real improvement, and worth spelling out because the hazard isn't obvious from the diff: this test mutates **shared module-level singletons** — it swaps the prototypes of `CommonMixin`, `DestroyMixin`, and `RadioMixin` and redefines `CommonMixin.__proto__`. Those objects are the ones every other spec in the run composes from. A failure that skipped restoration wouldn't fail *this* test loudly; it would corrupt everything after it, and the resulting cascade would point anywhere but here.

Three things the new version gets right:

Mutations now happen **inside** the `try`, with `mutatedMixins` recording what actually changed — so a throw partway through setup restores exactly what was applied rather than assuming all three succeeded. Restoration then walks that list in reverse.

`CommonMixin.__proto__` is captured as a **descriptor** beforehand and restored as one, with a `Reflect.deleteProperty` fallback that throws if the property can't be removed. Reassigning the value would have left it as a data property where it may not have been one.

`primaryFailed`/`cleanupFailed` are tracked separately and the primary error is rethrown first. Without that, a cleanup failure would mask the actual assertion failure — the classic way a `finally` block hides the thing you needed to see.

The assertions themselves are unchanged in substance: no `inheritedPollution` on either prototype, prototype identity still `Object.prototype`, and `__proto__` present as an own data property with the exact descriptor.

<sub>Reviewed with Claude Code.</sub>
